tree-building/tree_test.go

```diff
@@ -0,0 +1,305 @@
+			{Id: 3, Parent: 0},
+			{Id: 2, Parent: 0},
+			{Id: 1, Parent: 0},
+			{Id: 0},
+		},
+		expected: &Node{
+			Id: 0,
+			Children: []*Node{
+				{Id: 1},
+				{Id: 2},
+				{Id: 3},
+			},
+		},
+	},
+	{
+		name: "binary tree",
+		input: []Record{
+			{Id: 5, Parent: 1},
+			{Id: 3, Parent: 2},
+			{Id: 2, Parent: 0},
+			{Id: 4, Parent: 1},
+			{Id: 1, Parent: 0},
+			{Id: 0},
+			{Id: 6, Parent: 2},
+		},
+		expected: &Node{
+			Id: 0,
+			Children: []*Node{
+				{
+					Id: 1,
+					Children: []*Node{
+						{Id: 4},
+						{Id: 5},
+					},
+				},
+				{
+					Id: 2,
+					Children: []*Node{
+						{Id: 3},
+						{Id: 6},
+					},
+				},
+			},
+		},
+	},
+	{
+		name: "unbalanced tree",
+		input: []Record{
+			{Id: 5, Parent: 2},
+			{Id: 3, Parent: 2},
+			{Id: 2, Parent: 0},
+			{Id: 4, Parent: 1},
+			{Id: 1, Parent: 0},
+			{Id: 0},
+			{Id: 6, Parent: 2},
+		},
+		expected: &Node{
+			Id: 0,
+			Children: []*Node{
+				{
+					Id: 1,
+					Children: []*Node{
+						{Id: 4},
+					},
+				},
+				{
+					Id: 2,
+					Children: []*Node{
+						{Id: 3},
+						{Id: 5},
+						{Id: 6},
+					},
+				},
+			},
+		},
+	},
+}
+
+var failureTestCases = []struct {
+	name  string
+	input []Record
+}{
+	{
+		name: "root node has parent",
+		input: []Record{
+			{Id: 0, Parent: 1},
+			{Id: 1, Parent: 0},
+		},
+	},
+	{
+		name: "no root node",
+		input: []Record{
+			{Id: 1, Parent: 0},
+		},
+	},
+	{
+		name: "non-continuous",
+		input: []Record{
+			{Id: 2, Parent: 0},
+			{Id: 4, Parent: 2},
+			{Id: 1, Parent: 0},
+			{Id: 0},
+		},
+	},
+	{
+		name: "cycle directly",
+		input: []Record{
+			{Id: 5, Parent: 2},
+			{Id: 3, Parent: 2},
+			{Id: 2, Parent: 2},
+			{Id: 4, Parent: 1},
+			{Id: 1, Parent: 0},
+			{Id: 0},
+			{Id: 6, Parent: 3},
+		},
+	},
+	{
+		name: "cycle indirectly",
+		input: []Record{
+			{Id: 5, Parent: 2},
+			{Id: 3, Parent: 2},
+			{Id: 2, Parent: 6},
+			{Id: 4, Parent: 1},
+			{Id: 1, Parent: 0},
+			{Id: 0},
+			{Id: 6, Parent: 3},
+		},
+	},
+	{
+		name: "higher id parent of lower id",
+		input: []Record{
+			{Id: 0},
+			{Id: 2, Parent: 0},
+			{Id: 1, Parent: 2},
+		},
+	},
+}
+
+func (n Node) String() string {
+	return fmt.Sprintf("%d:%s", n.Id, n.Children)
+}
+
+func TestMakeTreeSuccess(t *testing.T) {
+	if TestVersion != testVersion {
+		t.Fatalf("Found TestVersion = %v, want %v", TestVersion, testVersion)
+	}
+	for _, tt := range successTestCases {
+		actual, err := Build(tt.input)
+		if err != nil {
+			t.Fatalf("Build for test case %q returned error %q.  Error not expected.",
+				tt.name, err)
+		}
+		if !reflect.DeepEqual(actual, tt.expected) {
+			t.Fatalf("Build for test case %q was expected to return\n%s\nbut returned:\n%s",
+				tt.name, tt.expected, actual)
+		}
+	}
+}
+
+func TestMakeTreeFailure(t *testing.T) {
+	for _, tt := range failureTestCases {
+		actual, err := Build(tt.input)
+		if err == nil {
+			t.Fatalf("Build for test case %q was expected to fail but returned: %s.",
```

Comment:
Agreed in the general case of a relatively short result message. For the special case of multiline return messages such as in ledger I think "Build was expected to return error but returned:\n%s" is better since it avoids having to add a line after the actual message.